pkg/p2p/kadcast/network.go

```diff
@@ -61,5 +63,61 @@ func sendUDPPacket(netw string, addr net.UDPAddr, payload []byte) {
 	if err != nil {
 		log.WithError(err).Warn("Error while writting to the filedescriptor.")
 		return
-	} 
-}
\ No newline at end of file
+	}
+}
+
+// StartTCPListener listens infinitely for TCP packet arrivals and
+// executes it's processing inside a gorutine by sending
+// the packets to the circularQueue.
+func StartTCPListener(netw string, queue *ring.Buffer, MyPeerInfo Peer) {
+	lAddr := getLocalTCPAddress()
+	// Set listening port.
+	lAddr.Port = int(MyPeerInfo.port)
+PacketConnCreation:
+	// listen to incoming TCP packets
+	listener, err := net.ListenTCP(netw, &lAddr)
+	if err != nil {
+		log.Panic(err)
+	}
+	// Set initial deadline.
+	listener.SetDeadline(time.Now().Add(time.Minute))
+
+	// Instanciate the buffer
+	buffer := make([]byte, 5000000)
+	for {
+		// Read TCP packet.
+		pc, err := listener.AcceptTCP()
+		uAddr := pc.RemoteAddr()
+		byteNum, err := io.ReadFull(pc, buffer)
+		if err != nil {
+			log.WithError(err).Warn("Error on packet read")
+			pc.Close()
+			goto PacketConnCreation
+		}
+		// Set a new deadline for the connection.
+		pc.SetDeadline(time.Now().Add(5 * time.Minute))
+		// Serialize the packet.
+		encodedPack := encodeReadTCPPacket(uint16(byteNum), uAddr, buffer[0:byteNum])
+		// Send the packet to the Consumer putting it on the queue.
+		queue.Put(encodedPack)
+		// Empty the buffer to read again.
+		buffer = make([]byte, 5000000)
+	}
+}
+
+// Opens a TCP connection with the peer sent on the params and transmits
+// a stream of bytes. Once transmited, closes the connection.
+func sendTCPStream(addr net.UDPAddr, payload []byte) {
+	conn, err := net.Dial("tcp", string(addr.IP)+":"+string(addr.Port))
+	if err != nil {
+		log.WithError(err).Warn("Could not stablish a connection with the dest Peer.")
+		return
+	}
+	defer conn.Close()
```

Comment:
Closing TCP connection after each Write would trigger TCP handshake/TCP teardown procedure on each sendTCPStream. Would it make sense to reuse the TCP connection or this is not applicable?

Reply:
I don't think you should leave a conn opened since we never consider that peers are avaliable slways.

Reply:
Not saying they are always available. In case a node does send more than once a message to the same peer connection, it should try to reuse the peer conn.

Usually conn.Close() could be called on Ping/Pong failure or write/read Timeout on the socket.

PS. Sorry if I'm missing anything from the protocol and talking too generally : )

Reply:
But assuming that a single node can have 20*128 peers at the same time. I don't think it is a good idea to keep the TCPconnections alive. 
They are opened, a CHUNK is sent, and closed, that's the process.

Reply:
If a node have 20*128 peers we can expect issues in other spheres due to overloading. 

Regarding the concern of keeping open 20*128 conn, with or without keeping tcp alive you'll probably need to exceed "fd limit per process" system parameter.